File: p445m/add_two_numbers2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
const List = require('list');
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  const [s1, s2] = [new Array(), new Array()];
  while (l1) {
    s1.push(l1.val);
    l1 = l1.next;
  }
  while (l2) {
    s2.push(l2.val);
    l2 = l2.next;
  }
  let [dummy, carry] = [new List.ListNode(0), 0];
  while (s1.length > 0 || s2.length > 0) {
    let v1 = s1.length > 0 ? s1.pop() : 0;
    let v2 = s2.length > 0 ? s2.pop() : 0;
    let sum = v1 + v2 + carry;
    p = new List.ListNode(sum % 10);
    p.next = dummy.next;
    carry = ~~(sum / 10);
    dummy.next = p;
  }
  if (carry > 0) {
    let p = new List.ListNode(carry);
    p.next = dummy.next;
    dummy.next = p;
  }
  return dummy.next;
};
// TEST
[
  [[], [], []],
  [[1], [], [1]],
  [[3], [5], [8]],
  [[6], [9], [1, 5]],
  [
    [6, 5],
    [9, 0, 2],
    [9, 6, 7],
  ],
  [
    [6, 5],
    [9, 4, 9],
    [1, 0, 1, 4],
  ],
  [
    [7, 2, 4, 3],
    [5, 6, 4],
    [7, 8, 0, 7],
  ],
  [[9, 9, 9], [1], [1, 0, 0, 0]],
].forEach(([a1, a2, expected]) => {
  const actual = List.toArray(
    addTwoNumbers(List.fromArray(a1), List.fromArray(a2))
  );
  console.log(a1, '+', a2, '=', actual);
  console.assert(actual.toString() === expected.toString());
});

File: p445m/add_two_numbers2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
from typing import List
from itertools import zip_longest
from local_packages.list import ListNode


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        s1, s2 = [], []
        while l1:
            s1.append(l1.val)
            l1 = l1.next
        while l2:
            s2.append(l2.val)
            l2 = l2.next
        dummy, carry = ListNode(), 0
        for e1, e2 in zip_longest(s1[::-1], s2[::-1]):
            s = carry
            if e1:
                s += e1
            if e2:
                s += e2
            carry = s // 10
            dummy.next = ListNode(s % 10, dummy.next)
        if carry == 1:
            dummy.next = ListNode(1, dummy.next)
        return dummy.next


# TESTS
for a1, a2, expected in [
    ([], [], []),
    ([1], [], [1]),
    ([3], [5], [8]),
    ([6], [9], [1, 5]),
    ([6, 5], [9, 0, 2], [9, 6, 7]),
    ([6, 5], [9, 4, 9], [1, 0, 1, 4]),
    ([7, 2, 4, 3], [5, 6, 4], [7, 8, 0, 7]),
    ([9, 9, 9], [1], [1, 0, 0, 0]),
]:
    sol = Solution()
    actual = ListNode.to_array(
        sol.addTwoNumbers(ListNode.from_array(a1), ListNode.from_array(a2))
    )
    print("Add", a1, "+", a2, "->", actual)
    assert actual == expected
